Question: Does age affect the ride duration for Citi Bike trips taken in January, 2016?

Alternative hypothesis: There is a negative correlation between age of riders and trip duration,
    with a .05 confidence level.
    
Null hypothesis: Age and trip duration are inversely related at less than a .05 confidence interval
    OR there is no apparent relationship
    OR there is a positive relationship.

In [1]:
from __future__  import print_function, division
import pylab as pl
import pandas as pd
import numpy as np
import os
import scipy

%pylab inline

if os.getenv ('PUI2016') is None:
    print ("Must set env variable PUI2016")
if os.getenv ('PUIDATA') is None:
    print ("Must set env variable PUI2016")

import os
import json
import seaborn

Populating the interactive namespace from numpy and matplotlib


In [2]:
get_ipython().system("curl -O 'https://s3.amazonaws.com/tripdata/201601-citibike-tripdata.zip'")

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 16.9M  100 16.9M    0     0  9903k      0  0:00:01  0:00:01 --:--:-- 9901k


In [3]:
get_ipython().system('unzip 201601-citibike-tripdata.zip')

Archive:  201601-citibike-tripdata.zip
  inflating: 201601-citibike-tripdata.csv  


In [4]:
df = pd.read_csv('201601-citibike-tripdata.csv')

In [5]:
df.head()

tripduration          starttime           stoptime  start station id  \
0           923  1/1/2016 00:00:41  1/1/2016 00:16:04               268   
1           379  1/1/2016 00:00:45  1/1/2016 00:07:04               476   
2           589  1/1/2016 00:00:48  1/1/2016 00:10:37               489   
3           889  1/1/2016 00:01:06  1/1/2016 00:15:56               268   
4          1480  1/1/2016 00:01:12  1/1/2016 00:25:52              2006   

       start station name  start station latitude  start station longitude  \
0   Howard St & Centre St               40.719105               -73.999733   
1         E 31 St & 3 Ave               40.743943               -73.979661   
2        10 Ave & W 28 St               40.750664               -74.001768   
3   Howard St & Centre St               40.719105               -73.999733   
4  Central Park S & 6 Ave               40.765909               -73.976342   

   end station id            end station name  end station latitude  \
0            3002  South End Ave & Liberty St             40.711512   
1             498          Broadway & W 32 St             40.748549   
2             284       Greenwich Ave & 8 Ave             40.739017   
3            3002  South End Ave & Liberty St             40.711512   
4            2006      Central Park S & 6 Ave             40.765909   

   end station longitude  bikeid    usertype  birth year  gender  
0             -74.015756   22285  Subscriber        1958       1  
1             -73.988084   17827  Subscriber        1969       1  
2             -74.002638   21997  Subscriber        1982       2  
3             -74.015756   22794  Subscriber        1961       2  
4             -73.976342   14562  Subscriber        1952       1

In [6]:
df.columns

Index(['tripduration', 'starttime', 'stoptime', 'start station id',
       'start station name', 'start station latitude',
       'start station longitude', 'end station id', 'end station name',
       'end station latitude', 'end station longitude', 'bikeid', 'usertype',
       'birth year', 'gender'],
      dtype='object')

In [7]:
df.drop([u'starttime', u'stoptime', u'start station id',
       u'start station name', u'start station latitude',
       u'start station longitude', u'end station id', u'end station name',
       u'end station latitude', u'end station longitude', u'bikeid',
       u'usertype', u'gender'], axis=1, inplace=True)

In [8]:
df.head()
# gender 1 is male, 2 is female

tripduration  birth year
0           923        1958
1           379        1969
2           589        1982
3           889        1961
4          1480        1952

### FBB: the following lines of code fail, but obviously the notebook is ot delivery ready 
### I am going to skip them  and go down to the cell where you were attepting to unzip. Look for cells that start with # FBB

In [9]:
#dropping NaN
df['age'].dropna(inplace= True)
df['tripduration'].dropna(inplace= True)

df['age'] = 2016 - df['birth year']
x = df['age']
y = df['tripduration']

KeyError: 'age'

In [ ]:
pl.scatter(x, y)

pl.xlabel('Age', fontsize = 18)
pl.ylabel('Trip Duration', fontsize = 18)
pl.title('Figure A: Relationship between age and trip duration of Citi Bike riders, January 2016')

I'm trying it next as a histogram instead of a scatter plot to see how it looks.

In [18]:
pl.hist([x, y], bins=np.arange(min(df), max(df) + 5, 5))

pl.xlabel('Age', fontsize = 18)
pl.ylabel('Trip Duration', fontsize = 18)

TypeError: float() argument must be a string or a number

In [34]:
scipy.stats.pearsonr(x, y)

(nan, 1.0)

In [36]:
len(x)

509478

In [37]:
len(y)

509478

In [46]:
array1 = x.dropna().values
len(array1)

484933

In [47]:
array2 = y.dropna().values
len(array2)

509478

In [51]:
array3 = np.random.choice(array2, 484933, replace = False)
len(array3)

484933

In [52]:
array1 = sorted(array1, key=int)

In [54]:
array3 = sorted(array3, key=int)

In [58]:
correlation = scipy.stats.pearsonr(array1, array3)
print (correlation)

(0.2004714664352909, 0.0)


# We cannot reject the null hypothesis. The Pearson test showed a correlation coefficient of .2 with a p-value of 0, meaning we can be 100% confident that the two variables have a slight positive correlation. As age goes up, so does trip duration.

The following four steps, an attempt to download data to PUIDATA, do not work. Did it locally instead. Will need to look into this later.

In [10]:
# FBB os.getenv should have been os.system, but also for consistency with earlier code you should stick to get_ipython()
# os.system("curl -O https://s3.amazonaws.com/tripdata/201507-citibike-tripdata.zip")
get_ipython().system("curl -O 'https://s3.amazonaws.com/tripdata/201507-citibike-tripdata.zip'")

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 32.9M  100 32.9M    0     0  10.0M      0  0:00:03  0:00:03 --:--:-- 10.0M


In [11]:
# FBB check if file is there
os.path.isfile("201507-citibike-tripdata.zip")

True

In [12]:
# FBB os.getenv should have been os.system
os.system("mv " + "201507-citibike-tripdata.zip " + os.getenv("PUIDATA"))

0

In [15]:
# FBB check it the files are there
os.path.isfile("201507-citibike-tripdata.zip"), os.path.isfile(os.getenv("PUIDATA") + "/201507-citibike-tripdata.zip")

(False, True)

In [18]:
# FBB this cell has 2 mistakes
os.getenv("unzip" + os.getenv("PUIDATA") + "/201507-citibike-tripdata.zip")

In [19]:
# FBB the argument of the function is wrong because youa are missing a space between unzip and os.getenv("PUIDATA")
# always print the string that you pass as arguments first to check if they should look right: it is easy to miss spaces
print("unzip" + os.getenv("PUIDATA") + "/201507-citibike-tripdata.zip")
print("unzip " + os.getenv("PUIDATA") + "/201507-citibike-tripdata.zip")

unzip/Users/fbianco/science/Dropbox//UI/PUIdata/201507-citibike-tripdata.zip


In [20]:
# FBB the command should be os.system not os.getenv!

os.system("unzip " + os.getenv("PUIDATA") + "/201507-citibike-tripdata.zip")

# FBB note that this command will unzip the .csv file  in the local directory, not in $PUIDATA. You should move it to 4PUIDATA
# but I would be sifficeintly happy with this amount of work and you reading the files locally at this point

0

In [21]:
# FBB make sure you undestand the commands os.system not os.getenv!
os.system?


In [22]:
# FBB make sure you undestand the commands os.system not os.getenv!
os.getenv?

In [14]:
# df2 = pd.read_csv('201507-citibike-tripdata.csv')
# df2.head()

In [98]:
get_ipython().system("curl -O 'https://s3.amazonaws.com/tripdata/201507-citibike-tripdata.zip'")

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 32.9M  100 32.9M    0     0  26.6M      0  0:00:01  0:00:01 --:--:-- 31.0M


In [100]:
get_ipython().system('unzip 201507-citibike-tripdata.zip')

Archive:  201507-citibike-tripdata.zip
  inflating: 201507-citibike-tripdata.csv  


In [101]:
df2 = pd.read_csv('201507-citibike-tripdata.csv')
df2.head()

tripduration          starttime           stoptime  start station id  \
0           732  7/1/2015 00:00:03  7/1/2015 00:12:16               489   
1           322  7/1/2015 00:00:06  7/1/2015 00:05:29               304   
2           790  7/1/2015 00:00:17  7/1/2015 00:13:28               447   
3          1228  7/1/2015 00:00:23  7/1/2015 00:20:51               490   
4          1383  7/1/2015 00:00:44  7/1/2015 00:23:48               327   

         start station name  start station latitude  start station longitude  \
0          10 Ave & W 28 St               40.750664               -74.001768   
1     Broadway & Battery Pl               40.704633               -74.013617   
2           8 Ave & W 52 St               40.763707               -73.985162   
3           8 Ave & W 33 St               40.751551               -73.993934   
4  Vesey Pl & River Terrace               40.715338               -74.016584   

   end station id               end station name  end station latitude  \
0             368             Carmine St & 6 Ave             40.730386   
1            3002     South End Ave & Liberty St             40.711512   
2             358  Christopher St & Greenwich St             40.732916   
3             250       Lafayette St & Jersey St             40.724561   
4              72               W 52 St & 11 Ave             40.767272   

   end station longitude  bikeid    usertype  birth year  gender  
0             -74.002150   18669  Subscriber      1970.0       1  
1             -74.015756   14618  Subscriber      1984.0       1  
2             -74.007114   18801  Subscriber      1992.0       1  
3             -73.995653   19137  Subscriber      1990.0       1  
4             -73.993929   15808  Subscriber      1988.0       1

In [105]:
df2.columns

Index([u'tripduration', u'starttime', u'stoptime', u'start station id',
       u'start station name', u'start station latitude',
       u'start station longitude', u'end station id', u'end station name',
       u'end station latitude', u'end station longitude', u'bikeid',
       u'usertype', u'birth year', u'gender'],
      dtype='object')

In [107]:
df2.drop([u'starttime', u'stoptime', u'start station id',
       u'start station name', u'start station latitude',
       u'start station longitude', u'end station id', u'end station name',
       u'end station latitude', u'end station longitude', u'bikeid',
       u'usertype', u'gender'], axis=1, inplace=True)

In [108]:
df2.head()

tripduration  birth year
0           732      1970.0
1           322      1984.0
2           790      1992.0
3          1228      1990.0
4          1383      1988.0

In [110]:
df2['age'] = 2015 - df2['birth year']
df2['age'].dropna(inplace= True)
df2['tripduration'].dropna(inplace= True)

x2 = df2['age']
y2 = df2['tripduration']

In [114]:
len(x2)

905311

In [115]:
len(y2)

1085676

In [118]:
arrayx2 = x2.dropna().values
arrayy2 = y2.dropna().values

In [119]:
len(arrayx2)

905311

In [120]:
len(arrayy2)

1085676

In [121]:
array4 = np.random.choice(arrayy2, 905311, replace = False)
len(array4)

905311

In [122]:
arrayx2 = sorted(arrayx2, key=int)
array4 = sorted(array4, key=int)

In [125]:
correlation = scipy.stats.pearsonr(arrayx2, array4)
print (correlation)

(0.14407973720036865, 0.0)


# These results back up those we got from the first sample. We have a correlation coefficient of .144 (similar to .2) and again, a p-value of 0. For Citi Bike riders in both January and July, age and trip duration have a slight positive correlation. The relationship is slightly weaker in the July 2015 data. The Null stands.